lib/Conversion/TorchOnnxToTorch/ComMicrosoftDomain.cpp

```diff
@@ -63,4 +63,393 @@ void mlir::torch::onnx_c::populateComMicrosoftDomain(
+                llvm::SmallVector<Value>{cstBatchSize, cstNumHeads,
+                                         cstSequenceLength, cstHeadSize});
+        Value qInput = rewriter.create<Torch::AtenReshapeOp>(
+            loc,
+            queryType.getWithSizesAndDtype(queryReshapeSizesInt,
+                                           queryType.getOptionalDtype()),
+            query, queryReshapeSizesList);
+
+        // Reshaping key.
+        SmallVector<int64_t, 6> kvReshapeSizesInt{batchSize, kvNumHeads,
+                                                  sequenceLength, headSize};
+        Value kvReshapeSizesList = rewriter.create<Torch::PrimListConstructOp>(
+            binder.getLoc(),
+            Torch::ListType::get(Torch::IntType::get(query.getContext())),
+            llvm::SmallVector<Value>{cstBatchSize, cstKVNumHeads,
+                                     cstSequenceLength, cstHeadSize});
+        Torch::ValueTensorType keyType =
+            cast<Torch::ValueTensorType>(key.getType());
+        Value kInput = rewriter.create<Torch::AtenReshapeOp>(
+            loc,
+            keyType.getWithSizesAndDtype(kvReshapeSizesInt,
+                                         keyType.getOptionalDtype()),
+            key, kvReshapeSizesList);
+
+        // Reshaping value.
+        Torch::ValueTensorType valueType =
+            cast<Torch::ValueTensorType>(value.getType());
+        Value vInput = rewriter.create<Torch::AtenReshapeOp>(
+            loc,
+            valueType.getWithSizesAndDtype(kvReshapeSizesInt,
+                                           valueType.getOptionalDtype()),
+            value, kvReshapeSizesList);
+
+        Value cstNone = rewriter.create<Torch::ConstantNoneOp>(loc);
+        Value cstFalse = rewriter.create<Torch::ConstantBoolOp>(loc, false);
+
+        Value qRotary = qInput, kRotary = kInput;
+        if (doRotary) {
+          // `totalSequenceLength` is a scalar tensor.
+          Value scalarTotalSeqLens = rewriter.create<Torch::AtenItemOp>(
+              loc, rewriter.getType<Torch::IntType>(), totalSequenceLength);
+          Value cstIntOne = rewriter.create<Torch::ConstantIntOp>(
+              binder.getLoc(), rewriter.getI64IntegerAttr(1));
+          Type boolTy = rewriter.getType<Torch::BoolType>();
+          Value condA = rewriter.create<Torch::AtenGtIntOp>(
+              loc, boolTy, cstSequenceLength, cstIntOne);
+          Value condB = rewriter.create<Torch::AtenNeIntOp>(
+              loc, boolTy, cstSequenceLength, scalarTotalSeqLens);
+          //   if (sequence_length > 1 && sequence_length !=
+          //   total_sequence_length)
+          //         is_subsequent_prompt = false;  // Subsequent prompt
+          Value isSubsequentPrompt = rewriter.create<Torch::Aten__And__BoolOp>(
+              loc, boolTy, condA, condB);
+
+          // Generating position_ids for rotary_embedding as follows:
+          //   pos_ids_a = torch.zeros((batch_size, seq_len), dtype=torch.int64)
```

Comment:
It's a complicated op. Better to provide a complete runnable pytorch script to demonstate the whole algorithm. 
Something like this https://colab.research.google.com/drive/1-Z4Mbfbt9v60-25sbSYmJusEJYtA8ItO?usp=sharing or [nonzero.py](https://gist.github.com/AmosLewis/f1b0dcad2c9ee2381f19d5836df952ff#file-nonzero-py)

Reply:
Hi @AmosLewis, the reference PyTorch implementation is https://gist.github.com/vivekkhandelwal1/f307b687fb133f36276f3d1a3c60ed7e. The code is just for the explanatory purpose, you may consider it as a pseudo code.